src/test/scala/com/microsoft/hyperspace/index/rules/JoinIndexRuleV2Test.scala

```diff
@@ -0,0 +1,247 @@
+    // Project [t2c1#4, t2c3#6]
+    //  +- Filter isnotnull(t2c1#4)
+    //   +- Relation[t2c1#4,t2c2#5,t2c3#6,t2c4#7] parquet
+
+    t3ProjectNode = Project(Seq(t3c1, t3c3), t3FilterNode)
+    // Project [t3c1#4, t3c3#6]
+    //  +- Filter isnotnull(t3c1#4)
+    //   +- Relation[t3c1#4,t3c2#5,t3c3#6,t3c4#7] parquet
+
+    createIndexLogEntry("t1i1", Seq(t1c1), Seq(t1c3), t1ProjectNode)
+    createIndexLogEntry("t1i2", Seq(t1c1, t1c2), Seq(t1c3), t1ProjectNode)
+    createIndexLogEntry("t1i3", Seq(t1c2), Seq(t1c3), t1ProjectNode)
+    createIndexLogEntry("t2i1", Seq(t2c1), Seq(t2c3), t2ProjectNode)
+    createIndexLogEntry("t2i2", Seq(t2c1, t2c2), Seq(t2c3), t2ProjectNode)
+    createIndexLogEntry("t3i1", Seq(t3c1), Seq(t3c3), t3ProjectNode)
+  }
+
+  before {
+    spark.conf.set(IndexConstants.INDEX_SYSTEM_PATH, systemPath.toUri.toString)
+    clearCache()
+  }
+
+  private def baseRelation(
+      location: FileIndex,
+      schema: StructType,
+      spark: SparkSession): HadoopFsRelation = {
+    HadoopFsRelation(location, new StructType(), schema, None, new ParquetFileFormat, Map.empty)(
+      spark)
+  }
+
+  override def afterAll(): Unit = {
+    FileUtils.delete(systemPath)
+    super.afterAll()
+  }
+
+  test("Join rule does not update plan for Broadcast Hash Join compatible joins") {
+    withSQLConf("spark.hyperspace.rule.joinV2.enabled" -> "true") {
+      val joinCondition = EqualTo(t1c1, t3c1)
+      val originalPlan =
+        Join(t1ProjectNode, t3ProjectNode, JoinType("inner"), Some(joinCondition))
+      val updatedPlan = JoinIndexRuleV2(originalPlan)
+      assert(updatedPlan.equals(originalPlan))
+    }
+  }
+
+  test("Join rule works if indexes exist and configs are set correctly") {
+    withSQLConf("spark.hyperspace.rule.joinV2.enabled" -> "true") {
+      val joinCondition = EqualTo(t1c1, t2c1)
+      val originalPlan =
+        Join(t1ProjectNode, t2ProjectNode, JoinType("inner"), Some(joinCondition))
+      val updatedPlan = JoinIndexRuleV2(originalPlan)
+      assert(!updatedPlan.equals(originalPlan))
+
+      val indexPaths = getIndexDataFilesPaths("t1i1") ++ getIndexDataFilesPaths("t2i1")
+      verifyUpdatedIndex(originalPlan, updatedPlan, indexPaths)
+    }
+  }
+
+  test("Join rule updates sort merge join part of a plan with both smj and bhj.") {
+    withSQLConf("spark.hyperspace.rule.joinV2.enabled" -> "true") {
+      val bhjCondition = EqualTo(t2c1, t3c1)
+      val bhjJoin = Join(t2ProjectNode, t3ProjectNode, JoinType("inner"), Some(bhjCondition))
+
+      val smjCondition = EqualTo(t1c1, t2c1)
+      val originalPlan = Join(t1ProjectNode, bhjJoin, JoinType("inner"), Some(smjCondition))
+
+      val updatedPlan = JoinIndexRuleV2(originalPlan)
+      assert(!updatedPlan.equals(originalPlan))
+
+      val indexPaths =
+        getIndexDataFilesPaths("t1i1") ++ getIndexDataFilesPaths("t2i1")
+      verifyUpdatedIndex(originalPlan, updatedPlan, indexPaths ++ Seq(new Path("t3")))
+    }
+  }
```

Comment:
Should we add some negative tests where index is not applied?

Reply:
I added one with no one-to-one matching of index columns. We have others where BHJ is not optimized. There is one test which I did not add : "index is not applied where no applicable index is found" but I thought that would not be necessary. Do you suggest we add this? Or any other case which I missed?